src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IAnalyzerHost.cs

```diff
@@ -24,10 +25,17 @@ public void AddAnalyzerAssembly(string analyzerAssemblyFullPath)
                 return;
             }
 
+            var fileChangeService = (IVsFileChangeEx)this.ServiceProvider.GetService(typeof(SVsFileChangeEx));
+            if (_visualStudioWorkspaceOpt == null)
+            {
+                // This can happen only in tests.
+                var testAnalyzer = new VisualStudioAnalyzer(analyzerAssemblyFullPath, fileChangeService, this.HostDiagnosticUpdateSource, this.Id, this.Workspace, loader: null, language: this.Language);
```

Comment:
There are lots of places in ProjectSystemShim testing where we can add more mocking and avoid null checks. AbstractProject already has checks for null workspace at lot of places. My goal with this change was to add enough mocking so we can test the interface APIs. For example, for AddAnalyzerReference, we want to test that we can add an analyzer reference and verify it was added, without necessary attempting to verify that diagnostic analyzers within it get correctly loaded up. The later would require mocking up all the analyzer related services, but doesn't gain us any additional testing surface for our newly added interface.